org.eclipse.xtext.web/src/main/java/org/eclipse/xtext/web/server/occurrences/OccurrencesService.java

```diff
@@ -0,0 +1,99 @@
+/**
+ * Copyright (c) 2015, 2020 itemis AG (http://www.itemis.eu) and others.
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ * 
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.eclipse.xtext.web.server.occurrences;
+
+import org.eclipse.emf.common.util.URI;
+import org.eclipse.emf.ecore.EObject;
+import org.eclipse.emf.ecore.EReference;
+import org.eclipse.xtext.EcoreUtil2;
+import org.eclipse.xtext.findReferences.IReferenceFinder;
+import org.eclipse.xtext.findReferences.TargetURIs;
+import org.eclipse.xtext.resource.ILocationInFileProvider;
+import org.eclipse.xtext.resource.IReferenceDescription;
+import org.eclipse.xtext.util.CancelIndicator;
+import org.eclipse.xtext.util.ITextRegion;
+import org.eclipse.xtext.util.ITextRegionWithLineInformation;
+import org.eclipse.xtext.util.TextRegion;
+import org.eclipse.xtext.util.concurrent.CancelableUnitOfWork;
+import org.eclipse.xtext.web.server.model.IXtextWebDocument;
+import org.eclipse.xtext.web.server.model.XtextWebDocumentAccess;
+import org.eclipse.xtext.web.server.util.CancelIndicatorProgressMonitor;
+import org.eclipse.xtext.web.server.util.ElementAtOffsetUtil;
+
+import com.google.common.base.Objects;
+import com.google.inject.Inject;
+import com.google.inject.Provider;
+import com.google.inject.Singleton;
+
+/**
+ * Service class for finding occurrences.
+ */
+@Singleton
+public class OccurrencesService {
+	@Inject
+	private ElementAtOffsetUtil elementAtOffsetUtil;
+
+	@Inject
+	private ILocationInFileProvider locationInFileProvider;
+
+	@Inject
+	private Provider<TargetURIs> targetURIsProvider;
+
+	@Inject
+	private IReferenceFinder referenceFinder;
+
+	/**
+	 * Find occurrences of the element at the given offset.
+	 */
+	public OccurrencesResult findOccurrences(XtextWebDocumentAccess document, int offset) {
+		return document.readOnly(new CancelableUnitOfWork<OccurrencesResult, IXtextWebDocument>() {
+			@Override
+			public OccurrencesResult exec(IXtextWebDocument doc, CancelIndicator cancelIndicator) throws Exception {
+				EObject element = OccurrencesService.this.elementAtOffsetUtil.getElementAt(doc.getResource(), offset);
```

Comment:
Remove `OccurrencesService.this.` everywhere

Reply:
why. this is a inner class and we refer to outer

Reply:
Hmm that doesn't work? gitpod lied to me

Reply:
will try again. can also be that jdt had a fuckup

Reply:
actually worked. wonder why not when i tried the first time